File: client/app/schemas/cards/choice_card.js
const { required } = require('../../modules/validations')
const { extend } = require('../../modules/utilities')
const cardSchema = require('../card')

module.exports = extend({}, cardSchema, {
  'data.body': {
    type: 'textarea',
    validations: [required],
  },
  'data.options': {
    type: 'list',
    validations: [required],
    columns: [
      {
        name: 'correct',
        type: 'select',
        options: [{ value: 'true' }, { value: 'false' }],
      },
      { name: 'value', type: 'text' },
      { name: 'feedback', type: 'text' },
    ],
  },
  'data.order': {
    type: 'select',
    validations: [required],
    options: [{ value: 'random' }, { value: 'set' }],
    default: 'random',
  },
  'data.max_options_to_show': {
    type: 'number',
    validations: [required],
    default: 4,
  },
})

File: server/schemas/cards/choice_card.py
from modules.content import get as c
from modules.util import extend, create_uuid_b64
from modules.validations import is_required, is_string, is_list, \
  is_one_of, is_boolean, is_integer, has_min_length
from schemas.card import schema as card_schema


def has_correct_options(options):
  """
  Ensure the list of options has at least one correct option.
  """

  has_correct = False
  for option in options:
    if option.get('correct') is True:
      has_correct = True
  if not has_correct:
    return c('error_need_correct')


schema = extend({}, card_schema, {
  'fields': {
    'data': {
      'embed': {
        'body': {  # Question field
          'validate': (is_required, is_string,),
        },
        'options': {  # Available answers
          'validate': (is_required, is_list, (has_min_length, 1), has_correct_options),
          'embed_many': {
            'id': {
              'validate': (is_required, is_string,),
              'default': create_uuid_b64,
            },
            'value': {
              'validate': (is_required, is_string,),
            },
            'correct': {
              'validate': (is_required, is_boolean,),
              'access': ('view',),
            },
            'feedback': {
              'validate': (is_required, is_string,),
              'access': ('view',),
            },
          }
        },
        'order': {
          'validate': (is_string, (is_one_of, 'random', 'set')),
          'default': 'random',
          'access': ('view',),
        },
        'max_options_to_show': {
          'validate': (is_integer,),
          'default': 4,
          'access': ('view',),
        },
      },
    },
  },
})
